src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarReader.cs

```diff
@@ -221,17 +221,13 @@ internal void AdvanceDataStreamIfNeeded()
                     return;
                 }
 
-                if (!dataStream.HasReachedEnd)
+                long remaining = dataStream.Remaining;
+                dataStream.SetReachedEnd();
```

Comment:
Rather than making the caller set end then advance could we make this all a method "AdvanceToEnd(Async)"?  That way we have a single method that's responsible for this and we don't have torn state between setting end and doing the seeking.